feat(linter): add linter service framework

HamdaanAliQuatil opened the pull request:
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint && make test`?
- [x] Have you added an entry to the changelog (`docs/reference/changelog.rst`)?

---

Closes #782 
This PR introduces a shared Linter Service to craft-application so downstream apps (Snapcraft, Charmcraft) can add and run linters consistently.

The design of the Linter Service is as discussed in #782

lengau commented on craft_application/commands/other.py:
```diff
@@ -29,6 +30,7 @@ def get_other_command_group() -> CommandGroup:
     """Return the lifecycle related command group."""
     commands: list[type[base.AppCommand]] = [
         InitCommand,
+        LintCommand,
```
Let's not make this a default command yet - apps should opt into this command since it only becomes useful once the app has at least one linter.

lengau commented on craft_application/services/linter.py:
```diff
@@ -0,0 +1,268 @@
+                self._issues.append(issue)
+                yield issue
+
+    def summary(self) -> ExitCode:
+        """Summarize results as an ExitCode based on highest severity."""
+        if any(i.severity == Severity.ERROR for i in self._issues):
+            return ExitCode.ERROR
+        if any(i.severity == Severity.WARNING for i in self._issues):
+            return ExitCode.WARN
+        return ExitCode.OK
```
This would probably be better as a `get_highest_severity` or similar. If we run linters during a lifecycle command (as we plan to), we don't want to exit nonzero on warnings.

lengau commented on craft_application/lint/types.py:
```diff
@@ -0,0 +1,110 @@
+    POST = "post"
+
+
+class Severity(str, Enum):
+    """Severity level for linter issues."""
+
+    INFO = "INFO"
+    WARNING = "WARNING"
+    ERROR = "ERROR"
```
Let's make these levels comparable so I can say:

```python
Severity.ERROR > Severity.WARNING
Severity.ERROR >= Severity.INFO
Severity.WARNING > Severity.INFO
```

lengau commented on craft_application/services/linter.py:
```diff
@@ -0,0 +1,268 @@
+    def register(cls, linter_cls: type[AbstractLinter]) -> None:
+        """Register a linter class for use by the service."""
+        if inspect.isabstract(linter_cls):
+            return
```
Surely this should raise an exception instead?

lengau commented on craft_application/services/linter.py:
```diff
@@ -0,0 +1,268 @@
+        """Register a linter class for use by the service."""
+        if inspect.isabstract(linter_cls):
+            return
+        stage = getattr(linter_cls, "stage", None)
+        name = getattr(linter_cls, "name", None)
```
This is something that `AbstractLinter` should be checking, not the linter service.

lengau commented on craft_application/services/linter.py:
```diff
@@ -0,0 +1,268 @@
+        """App-specific filtering hook."""
+        yield from issues
+
+    def run(
+        self,
+        stage: Stage,
+        ctx: LintContext,
+    ) -> Iterator[LinterIssue]:
```
We should have an integration or end-to-end test there that shows a linter erroring out and then an update to the project having it ignored.

lengau reviewed (dismissed):
Thanks! I've done an initial review with a few things that need changing.

medubelko commented:
Hi @HamdaanAliQuatil, thanks for taking this on. :)

I'm holding off my documentation review until the implementation is complete. If it looks like the implementation is a ways off, it might be better to split the documentation portion into its own PR afterward.

lengau commented on craft_application/commands/lint.py:
```diff
@@ -0,0 +1,101 @@
+            cli_ignore_files=list(parsed_args.lint_ignore_files or []),
+        )
+        for _ in linter.run(Stage(parsed_args.stage), ctx):
+            pass
```
This should output the info/warning/errors for each linter that got run.

lengau reviewed (commented):
Looking good so far!

Please add a couple of linters to `testcraft` and a spread test that tests for the app outputting the expected values.

A good pre linter would be a recommended field (e.g. `version`) being missing. A good post linter would be that the artifact is empty other than the generated metadata.yaml file.

bepri commented on docs/reference/services/linter.rst:
```diff
@@ -0,0 +1,124 @@
+
+- ``@classmethod register(linter_cls)``
+- ``@classmethod build_ignore_config(project_dir, cli_ignores=None, cli_ignore_files=None)``
+  â€“ merge defaults â†’ files â†’ CLI (CLI takes precedence); intended to be
```
There's some weird character encoding going on in this section

bepri commented on docs/reference/services/linter.rst:
```diff
@@ -0,0 +1,124 @@
+- Central ignore rules: the service owns ``IgnoreConfig`` and enforces user
+  intent via ``should_ignore`` before applying app-specific policy hooks.
+
+API
```
I think a lot of this section can be turned into [an autodoc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass). As an example, see: https://github.com/canonical/craft-application/blob/3d31a9fa7a58828043b555f6f39cc7561cd10917/docs/reference/services/project.rst?plain=1#L98-L101

bepri commented on craft_application/services/linter.py:
```diff
@@ -0,0 +1,278 @@
+            cls._merge_into(config, file_cfg)
+        if cli_ignore_files:
+            for extra in cli_ignore_files:
+                if extra and extra.exists():
```
If I understand correctly, `extra` should exist as it was specified by the user? I would raise an error instead of silently ignoring it, they might've made a typo and will be confused when their lints aren't ignored.

bepri commented on craft_application/services/linter.py:
```diff
@@ -0,0 +1,278 @@
+        cli_ignores: list[str] | None = None,
+        cli_ignore_files: list[Path] | None = None,
+    ) -> IgnoreConfig:
+        """Load and cache ignore config, merging CLI rules if any."""
```
I'm not sure about calling this a "cache" - calling `load_ignore_config` multiple times from a downstream application will still call `build_ignore_config` multiple times.

bepri commented on testcraft/services/linter.py:
```diff
@@ -0,0 +1,108 @@
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+PROJECT_FILE = "testcraft.yaml"
```
```suggestion
from testcraft.application import TESTCRAFT
PROJECT_FILE = f"{TESTCRAFT.name}.yaml"
```

bepri commented on craft_application/util/logging.py:
```diff
@@ -48,7 +48,7 @@ def handle_runtime_error(
     error: BaseException,
     *,
-    print_error: Callable[[craft_cli.CraftError], None] = craft_cli.emit.report_error,
+    print_error: Callable[[craft_cli.CraftError], None] = craft_cli.emit.error,
```
Why is this change here?

bepri reviewed (commented):
Thanks! I'm down to just nitpicks now, personally.

bepri commented on tests/unit/services/test_linter_service.py:
```diff
@@ -0,0 +1,113 @@
+    assert issues == []
+    assert svc.issues_by_linter == {}
+    assert int(svc.summary()) == 0  # OK
+
```
Maybe add the registration as part of the autouse fixture?

HamdaanAliQuatil commented:
@bepri could you please retrigger the tests

HamdaanAliQuatil commented:
noted 🫡 @bepri I've reverted the missed changes and fixed the spread test failure to ensure the Testcraft linter service correctly extracts .testcraft tarballs even when no artifact directories are specified. This allows to fire TC100 again, and I've added a unit test to cover this behavior.

bepri commented on craft_application/commands/lint.py:
```diff
@@ -0,0 +1,162 @@
+        ctx = LintContext(project_dir=project_dir, artifact_dirs=artifact_dirs)
+
+        linter = services.get("linter")
+        cli_ignore_rules = list(parsed_args.lint_ignores or [])
```
This, and the line for `cli_ignore_files` below, are unnecessary list conversions since argparse is already giving us a list:

```py
>>> parser = argparse.ArgumentParser()
>>> _ = parser.add_argument("-a", action="append", type=str, dest="a")
>>> args = parser.parse_args(["-a", "1", "-a", "2", "-a", "3"])
>>> args.a
['1', '2', '3']
>>> type(args.a)
<class 'list'>
```

Instead, you should cast it with `typing.cast`

bepri reviewed (approved):
I'm approving this because I'm happy with the current functionality, but I have just a couple more minor nitpicks about the code that I'd like to be added before this is merged. Thanks!

@medubelko @lengau I think this is ready for your review.

medubelko commented on craft_application/commands/lint.py:
```diff
@@ -0,0 +1,162 @@
+            emit.message("lint: ERROR")
+            return 2
+
+        emit.message("lint: WARN")
```
```suggestion
        if highest is None:
            emit.message(f"Linted {filename} successfully.")
            return 0
        if highest.name == "ERROR":
            emit.message(f"Errors found in {filename}")
            return 2

        emit.message(f"Possible issues found in {filename}")
```

This is more in line with our CLI commands.

medubelko commented on craft_application/commands/lint.py:
```diff
@@ -0,0 +1,162 @@
+                    location = f" ({issue.filename})" if issue.filename else ""
+                    emit.message(
+                        f"  - [{issue.severity.name}] {issue.id}: {issue.message}{location}"
+                    )
```
From some initial testing, it seems that YAML syntax errors never reach this code. Instead, `YamlError` (I believe)  interrupts. Is this expected?

Source:

```yaml
name: testcraft-test
version: "1.0"

base: ubuntu@24.04
platforms:
  platform-independent:
    build-on: [amd64, arm64, ppc64el, s390x, riscv64]
    build-for: [all]

parts:
  my-test:
    plugin: nil

parts:
```

Result:

```bash
$ testcraft lint
error parsing '(unknown)': found duplicate key 'parts'
Detailed information: 
while constructing a mapping
found duplicate key 'parts'
  in "<unicode string>", line 1, column 1:
    name: testcraft-test
    ^
Recommended resolution: Ensure (unknown) contains valid YAML
```

medubelko reviewed (commented):
Hi @HamdaanAliQuatil, I'm a technical author at Canonical who looks at design and documentation.

I did some testing, and I have some questions about the UX and a suggestion for the CLI messages. Let me know what you think.

lengau commented on docs/reference/services/linter.rst:
```diff
@@ -0,0 +1,123 @@
+   :show-inheritance:
+
+How to add a new linter
+-----------------------
```
This should be expanded and added as its own how-to guide.

lengau commented on craft_application/commands/lint.py:
```diff
@@ -0,0 +1,163 @@
+            default=[],
+            metavar="PATH",
+            help=("Path to YAML ignore file. May repeat. CLI rules take precedence."),
+        )
```
I don't think we want this in a separate file. It'll probably go into the `*craft.yaml` file, but that's still undecided, so for right now could you please remove this?

HamdaanAliQuatil commented:
Hi @mr-cal - yes, still working on this. Sorry for the radio silence; work got busy. I’m back on it now and will push the remaining changes. I’ll post an update by eod.

mr-cal commented:
@HamdaanAliQuatil, I'm agreed with all of @lengau's responses and suggestions to my review.

HamdaanAliQuatil commented on craft_application/commands/lint.py:
```diff
@@ -0,0 +1,155 @@
+            ),
+        )
+
+    def run(self, parsed_args: argparse.Namespace) -> int | None:
```
alright. ill follow-up on this in a separate discussion. removing from `craft_application` and adding it to `testcraft` for now.

HamdaanAliQuatil commented:
@lengau @mr-cal i've updated the PR

mr-cal reviewed (commented):
Aside from the failing tests, I'm happy with this.

mr-cal commented on docs/how-to-guides/add-a-linter.rst:
```diff
@@ -0,0 +1,82 @@
+   svc.load_ignore_config(project_dir=Path.cwd())
+   for issue in svc.run(Stage.PRE, LintContext(Path.cwd(), [])):
+       print(issue)
+   print(int(svc.summary()))  # 0 OK, 1 WARN, 2 ERROR
```
I think copilot is confused.

medubelko commented on docs/how-to-guides/add-a-linter.rst:
```diff
@@ -0,0 +1,82 @@
+1) Implement a linter
+---------------------
+
+Create a class that extends ``AbstractLinter`` and yields issues from ``run``.
```
```suggestion
Create a new class that extends the ``AbstractLinter`` class and yields issues in its ``run()`` method.
```

medubelko commented on docs/how-to-guides/add-a-linter.rst:
```diff
@@ -0,0 +1,82 @@
+``ctx.project``) and ``Stage.POST`` to lint packed artifacts via
+``ctx.artifact_dirs``.
+
+.. code-block:: python
```
Please provide an example filename with:

```
    :caption: <path/to/file>.py
```

medubelko commented on docs/how-to-guides/add-a-linter.rst:
```diff
@@ -0,0 +1,82 @@
+   print(int(svc.summary()))  # 0 OK, 1 WARN, 2 ERROR
+
+4) Ignore configuration
+-----------------------
```
```suggestion
Linter ignores by the user
--------------------------
```

medubelko commented on docs/how-to-guides/add-a-linter.rst:
```diff
@@ -0,0 +1,168 @@
+* Iterate every directory in ``ctx.artifact_dirs`` instead of assuming a single artifact.
+* Validate packaged content (for example, required files, layout, or metadata).
+* Keep checks read-only and deterministic (avoid mutating files or relying on network access).
+
```
```suggestion

Here's an example linter that checks project and compiled part files:
```
